examples/siro_sandbox/server/remote_gui_input.py

```diff
@@ -0,0 +1,262 @@
+# LICENSE file in the root directory of this source tree.
+
+import magnum as mn
+from server.average_rate_tracker import AverageRateTracker
+
+from habitat.gui.gui_input import GuiInput
+
+
+class RemoteGuiInput:
+    def __init__(self, interprocess_record, debug_line_render):
+        self._recent_client_states = []
+        self._interprocess_record = interprocess_record
+        self._debug_line_render = debug_line_render
+
+        self._receive_rate_tracker = AverageRateTracker(2.0)
+
+        # temp hack
+        self._key_map = {
+            "q": GuiInput.KeyNS.Q,
+            "w": GuiInput.KeyNS.W,
+            "e": GuiInput.KeyNS.E,
+            "a": GuiInput.KeyNS.A,
+            "s": GuiInput.KeyNS.S,
+            "d": GuiInput.KeyNS.D,
+            "f": GuiInput.KeyNS.F,
+        }
+
+        # temp map VR button to key
+        self._button_map = {
+            0: GuiInput.KeyNS.ZERO,
+            1: GuiInput.KeyNS.ONE,
+            2: GuiInput.KeyNS.TWO,
+            3: GuiInput.KeyNS.THREE,
+        }
+
+        self._gui_input = GuiInput()
+
+    def get_gui_input(self):
+        return self._gui_input
+
+    def get_history_length(self):
+        return 4
+
+    def get_history_timestep(self):
+        return 1 / 60
+
+    def get_head_pose(self, history_index=0):
+        if history_index >= len(self._recent_client_states):
+            return None, None
+
+        avatar_root_json = self._recent_client_states[history_index]["avatar"][
+            "root"
+        ]
+        pos_json = avatar_root_json["position"]
+        pos = mn.Vector3(pos_json[0], pos_json[1], pos_json[2])
+        rot_json = avatar_root_json["rotation"]
+        rot_quat = mn.Quaternion(
+            mn.Vector3(rot_json[1], rot_json[2], rot_json[3]), rot_json[0]
+        )
+
+        return pos, rot_quat
+
+    def get_hand_pose(self, hand_idx, history_index=0):
+        if history_index >= len(self._recent_client_states):
+            return None, None
+
+        client_state = self._recent_client_states[history_index]
+        assert "hands" in client_state["avatar"]
+        hands_json = client_state["avatar"]["hands"]
+        assert hand_idx >= 0 and hand_idx < len(hands_json)
+
+        hand_json = hands_json[hand_idx]
+        pos_json = hand_json["position"]
+        # pos = mn.Vector3(pos_json["_x"], pos_json["_y"], pos_json["_z"])
+        pos = mn.Vector3(pos_json[0], pos_json[1], pos_json[2])
+        rot_json = hand_json["rotation"]
+        # rot_quat = mn.Quaternion(mn.Vector3(rot_json["_x"], rot_json["_y"], rot_json["_z"]), rot_json["_w"])
+        rot_quat = mn.Quaternion(
+            mn.Vector3(rot_json[1], rot_json[2], rot_json[3]), rot_json[0]
+        )
+
+        return pos, rot_quat
+
+    def update_input_state_from_remote_client_states(self, client_states):
+        if not len(client_states):
+            return
+
+        # gather all recent keyDown and keyUp events
+        for client_state in client_states:
```

Comment:
Conceptually, we should expect several `client_states` to be consolidated into a single `GuiInput` state to be processed on a single server frame. E.g., button-presses and releases over several frames (on the client) get consolidated to a bunch of button-presses on a single frame.

We should expect client-side GuiInput to be recorded at a high rate, e.g. 60+ FPS, so that we can get high-fidelity headset- and controller-tracking.